File: Day5Part2.java
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.stream.Collectors;
import java.util.List;
import java.util.Collections;
import java.util.Optional;

public class Day5Part2 {
    public static void main(String[] args) throws Exception {
        List<Integer> seatIds = Files.lines(Paths.get("input-day5.txt"))
            .map(Seat::from)
            .map(Seat::id)
            .collect(Collectors.toList());

        Collections.sort(seatIds);

        findMySeat(seatIds).ifPresent(System.out::println);
    }

    static Optional<Integer> findMySeat(List<Integer> seatIds) {
        if (seatIds.size() < 2) return Optional.empty();
        int a = seatIds.get(0) + 1;
        int b = seatIds.get(1);
        return a != b ? Optional.of(a) : findMySeat(seatIds.subList(1, seatIds.size()));
    }

    static class Seat {
        int row;
        int col;
        int id() { return row * 8 + col; }

        private static int pow2(int exponent) { return exponent == 0 ? 1 : 2 * pow2(--exponent); }

        private static int parse(String s, int min, int max, int exponent, char upper) {
            if (s.isEmpty()) return min;
            int delta = pow2(exponent--);
            return (s.charAt(0) == upper)
                ? parse(s.substring(1), min + delta, max,         exponent, upper)
                : parse(s.substring(1), min,         max - delta, exponent, upper);
        }

        private static int parseCol(String s) {
            return parse(s, 0, 7, 2, 'R');
        }

        private static int parseRow(String s) {
            return parse(s, 0, 127, 6, 'B');
        }

        Seat(int row, int col) { 
            this.row = row; 
            this.col = col; 
        }

        static Seat from(String s) {
            int row = parseRow(s.substring(0,7));
            int col = parseCol(s.substring(7));
            return new Seat(row, col);
        }

        public String toString() {
            return String.format("row %d, column %d, seat ID %d", row, col, id());
        }
    }
}

File: day5part2.clj
(defn pow2
  [n]
  (if (= 0 n)
    1
    (* 2 (pow2 (dec n)))))
 
(defn parse 
  [s mn mx expo upchar]
  (if (empty? s)
    mn
    (let [rs (apply str (rest s))
          delta (pow2 expo)]
      (if (= upchar (first s))
        (parse rs (+ mn delta) mx          (dec expo) upchar)
        (parse rs mn           (- mx delta)(dec expo) upchar)))))

(defn id [[r c]] (+ c (* r 8)))
(defn parse-row [s] (parse s 0 127 6 \B))
(defn parse-col [s] (parse s 0 8 2 \R))
(defn parse-seat 
  [s]
  (list 
    (parse-row (subs s 0 7))
    (parse-col (subs s 7 10))))

(defn find-my-seat
  [ids]
  (if (> 2 (count ids))
    nil
    (let [a (inc (first ids))
          b (second ids)]
      (if (not= a b)
        a
        (find-my-seat (rest ids))))))

(let [seat-ids (sort
                 (map (comp id parse-seat)
                     (.split (slurp "input-day5.txt") "\n")))]
  (println (find-my-seat seat-ids)))

